clientlibrary/worker/shard-consumer.go

```diff
@@ -240,22 +233,39 @@ func (sc *ShardConsumer) getRecords(shard *par.ShardStatus) error {
 		// The shard has been closed, so no new records can be read from it
 		if getResp.NextShardIterator == nil {
 			log.Infof("Shard %s closed", shard.ID)
-			shutdownInput := &kcl.ShutdownInput{ShutdownReason: kcl.TERMINATE, Checkpointer: recordCheckpointer}
-			sc.recordProcessor.Shutdown(shutdownInput)
+			sc.shutdown(kcl.TERMINATE)
+
 			return nil
 		}
 		shardIterator = getResp.NextShardIterator
 
-		select {
-		case <-*sc.stop:
```

Comment:
Need to understand why this was removed

Reply:
ah this was one of the red flags lol... so the `*sc.stop` channel is the same channel the worker code uses for shutting down. so there were multiple things consuming from the same channel, basically causing a race and things were not shutting down correctly.